include/GoogleSync/GoogleSync.php

```diff
@@ -0,0 +1,1167 @@
+                $foundOnCalendar = $calendarListEntry->getId();
+                break;
+            }
+        }
+        if (isset($foundOnCalendar)) {
+            $GLOBALS['log']->info(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Found Event on Calendar ID: ' . $foundOnCalendar);
+            $result = $this->gService->events->move($foundOnCalendar, $event_id, $this->calendarId);
+            return $result;
+        } else {
+            $GLOBALS['log']->warn(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Unable to find missing Calendar Event');
+            return null;
+        }
+    }
+
+    /**
+     * Get a SuiteCRM meeting by Google Event ID
+     *
+     * @param string $event_id The Google Event ID
+     *
+     * @return \Meeting SuiteCRM Meeting Bean
+     */
+    public function getMeetingByEventId($event_id)
+    {
+
+        // We do it this way so we also get deleted meetings
+        $query = "SELECT id FROM meetings WHERE gsync_id = '{$event_id}'";
+        $result = $this->db->query($query);
+
+        // This checks to make sure we only get one result. If we get more than one, something is inconsistant in the DB
+        if ($result->num_rows > 1) {
+            $GLOBALS['log']->fatal(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'More than one meeting matches Google Id: ' . $event_id);
+            return false;
+        } elseif ($result->num_rows == 0) {
+            return; // No matches Found
+        }
+
+        $meeting = new Meeting;
+        $row = $this->db->fetchByAssoc($result);
+        $meeting->retrieve($row['id'], true, false);
+
+        return $meeting;
+    }
+
+    /**
+     * Creates and Sets $this->gService to a valid Google Calendar Service
+     *
+     * @return bool Success/Failure
+     */
+    public function setGService()
+    {
+        // make sure we have a client set
+        if (!isset($this->gClient)) {
+            $GLOBALS['log']->fatal(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'The Google Client is not set up. See setClient Method');
+            return false;
+        }
+
+        // create new calendar service
+        $this->gService = new Google_Service_Calendar($this->gClient);
+        if (isset($this->gService)) {
+            return true;
+        } else {
+            $GLOBALS['log']->fatal(__FILE__ . ':' . __LINE__ . ' ' . __METHOD__ . ' - ' . 'Setting $this->gService Failed');
+            return false;
+        }
+    }
+
+    /**
+     * Figure out if we need to push/pull an update, or do nothing.
+     *
+     * Used when an event w/ a matching ID is on both ends of the sync.
+     *
+     * @param Meeting|\Google_Service_Calendar_Event $event_1 Either the Meeting Bean or Google_Service_Calendar_Event Object
+     * @param \Google_Service_Calendar_Event $event_2 (optional) Google Event
+     *
+     * @return string push(_delete), pull(_delete), skip
+     */
+    public function pushPullSkip($event_1, $event_2 = null)
```

Comment:
This is by far the most complicated method in the class, but for good reason. It needs to take two objects, determine which is which, then figure out what event to sync in what direction and whether to delete one of them to match the other. It then returns a simple string with instructions for the next step of the process.
I'd love to simplify it, but I can't find any basic way to do that without complicating something else.
One possibility is to split it into two methods, but the split off method would only ever be called by the original. It's useless without it. I don't think it would actually be simpler, just more difficult to read and follow.
I'm not sure where to go with this. :\

Reply:
> One possibility is to split it into two methods, but the split off method would only ever be called by the original. It's useless without it.
Yes, spliting it into two method would be the correct solution.

> I don't think it would actually be simpler, just more difficult to read and follow.
Then, please think about the tests. Each complex function makes more difficult to write proper test to cover the method.